### PR TITLE
fix: implement worker invocation cancellation

### LINKED ISSUE #334: [Enhancement]: Implement worker invocation cancellation across Rust and Python SDKs
### Affected area

- Rust core runtime
- Plugins
- Python worker SDK
- Documentation and examples

### Problem or opportunity

The `grpc-v1` worker protocol reserves an explicit cancellation operation and assigns every callback invocation an `invocation_id`:

- [`PluginWorker.CancelInvocation`](https://github.com/NVIDIA/NeMo-Relay/blob/4b9a5307e7efac98d9e57193dc8c5141bb5d4dec/crates/worker-proto/proto/nemo/relay/worker/v1/plugin_worker.proto#L8-L16)
- [`InvokeRequest.invocation_id` and `CancelInvocationRequest`](https://github.com/NVIDIA/NeMo-Relay/blob/4b9a5307e7efac98d9e57193dc8c5141bb5d4dec/crates/worker-proto/proto/nemo/relay/worker/v1/plugin_worker.proto#L134-L141)

Cancellation is not implemented end to end today:

- The [Rust worker SDK returns `accepted = false`](https://github.com/NVIDIA/NeMo-Relay/blob/4b9a5307e7efac98d9e57193dc8c5141bb5d4dec/crates/worker/src/lib.rs#L974-L984).
- The Python worker SDK introduced in PR #310 likewise [returns `accepted = false`](https://github.com/NVIDIA/NeMo-Relay/blob/7dea73e4b36de638d840194ad32da8a9d050a8e1/python/plugin/src/nemo_relay_plugin/_api.py#L1247-L1249).
- The Rust host creates invocation IDs but does not currently send `CancelInvocation`. Unary worker calls instead pass through a [fixed 30-second RPC timeout](https://github.com/NVIDIA/NeMo-Relay/blob/18e926b4c1349aa8e3e3181ae0366a8b479ef9d9/crates/core/src/plugin/dynamic/worker.rs#L72-L78).

Transport-level cancellation may drop an async handler future in some cases, but the worker contract does not currently track callback tasks by `invocation_id` or guarantee cancellation behavior. In particular, synchronous Python callbacks cannot be preempted by cancelling an asyncio task. Long-running callbacks may therefore continue work or side effects after the host caller has timed out or stopped consuming a stream, while host-owned continuations and invocation scope mappings are being removed.

This limitation was identified and accepted as a follow-up in [PR #310's review thread](https://github.com/NVIDIA/NeMo-Relay/pull/310#discussion_r3499165293); see also the [author response](https://github.com/NVIDIA/NeMo-Relay/pull/310#discussion_r3499237015).

### Proposed enhancement

Define and implement invocation cancellation across the Relay worker host, Rust worker SDK, and Python worker SDK:

1. Track active unary and streaming invocations by `invocation_id` in each worker SDK.
2. Have the host send `CancelInvocation` when a managed call is cancelled, a worker invocation times out, or a worker-backed stream is abandoned when explicit cancellation is required.
3. Cancel the corresponding Rust future/task or Python asyncio task cooperatively and return a meaningful `WorkerAck`.
4. Make repeated cancellation and cancellation racing with normal completion safe and deterministic.
5. Coordinate continuation and scope-stack cleanup so cancellation cannot leave stale host state or allow continuation use after the invocation ends.
6. Represent cancellation distinctly from generic worker failure in host-facing errors and stream termination.
7. Decide whether cancellation is mandatory for `grpc-v1` or negotiated as a worker capability so older SDKs remain compatible.

Synchronous callbacks that cannot be safely preempted should have explicit documented behavior. The contract should not claim that a cancellation acknowledgment guarantees interruption of arbitrary blocking user code.

### Runtime contract and binding impact

- **Rust core host:** propagate caller cancellation and timeout into the worker protocol; own final continuation and scope cleanup.
- **Rust worker SDK:** maintain active invocation handles and cancel asynchronous handler work by invocation ID.
- **Python worker SDK:** maintain active `asyncio.Task` instances and cancel asynchronous handlers by invocation ID; document the limitation for synchronous blocking callbacks.
- **Other worker SDKs:** future worker implementations should follow the same `grpc-v1` cancellation semantics or advertise that the capability is unsupported.
- **Native/in-process plugins:** not directly affected because they do not cross the worker RPC boundary, although managed-call cancellation semantics should remain consistent where possible.

### Alternatives considered

- **Rely only on the 30-second host timeout:** this bounds how long the host waits for unary RPCs but does not define worker-side cancellation, synchronous callback behavior, or cleanup semantics.
- **Rely only on gRPC transport cancellation:** transport cancellation is useful but does not replace an explicit cross-language contract keyed by `invocation_id`, especially for blocking callbacks and coordinated host state cleanup.
- **Document cancellation as unsupported:** acceptable as a short-term beta limitation, but incomplete for long-running execution intercepts and worker processes.
- **Terminate the entire worker process:** provides a coarse stop mechanism but disrupts unrelated concurrent invocations and loses plugin state.

### Acceptance criteria

- [ ] The host invokes `CancelInvocation` for the defined cancellation and timeout paths.
- [ ] Rust and Python worker SDKs track active unary and streaming invocations by `invocation_id`.
- [ ] Active asynchronous callbacks can be cancelled, and workers return an accurate acknowledgment.
- [ ] Cancellation of unknown, completed, or already-cancelled invocation IDs has documented, idempotent behavior.
- [ ] Continuations, invocation scope mappings, and stream resources are released exactly once across completion/cancellation races.
- [ ] Rust host and SDK tests cover timeout, caller cancellation, streaming cancellation, repeated cancellation, and completion races.
- [ ] Python SDK tests cover asynchronous callback cancellation and the documented behavior of synchronous callbacks.
- [ ] At least one end-to-end worker integration test verifies host-to-worker cancellation.
- [ ] `grpc-v1` compatibility or capability-negotiation behavior is defined for workers that do not support cancellation.
- [ ] Worker authoring documentation explains cancellation guarantees and limitations.

### References

- Relates to #310
- [PR #310 cancellation discussion](https://github.com/NVIDIA/NeMo-Relay/pull/310#discussion_r3499165293)
- [Rust worker SDK cancellation stub](https://github.com/NVIDIA/NeMo-Relay/blob/4b9a5307e7efac98d9e57193dc8c5141bb5d4dec/crates/worker/src/lib.rs#L974-L984)

### PR DESCRIPTION
**Pull request by willkill07**
#### Overview

Implement end-to-end cooperative cancellation for `grpc-v1` worker invocations across the Relay host and the Rust and Python worker SDKs.

- [x] I confirm this contribution is my own work, or I have the right to submit it under this project's license.
- [x] I searched existing issues and open pull requests, and this does not duplicate existing work.

#### Details

- Track active unary and streaming callbacks by `invocation_id` in both worker SDKs.
- Send `CancelInvocation` when a host caller is dropped, a worker RPC times out, or a worker-backed stream is abandoned.
- Make cancellation idempotent and race-safe, release continuation and scope state exactly once, and surface cancellation as `worker.cancelled`.
- Document cooperative cancellation guarantees, including the limitation for blocking Python callbacks.
- Add Rust host/unit/integration coverage plus Rust and Python SDK tests for timeout, caller/stream cancellation, repeated requests, and completion races.
- Preserve compatibility with older `grpc-v1` workers that return `accepted = false`.

Validation:
- Full Rust, Python, Go, Node.js, WebAssembly, and documentation validation matrix.
- `uv run pre-commit run --all-files`.
- Commit-time Ruff, type, docs-link, Cargo formatting, Clippy, and Cargo check hooks.
- The Python matrix excluded the known pre-existing macOS hang in `test_unlink_unix_socket_refuses_an_active_socket` after reproducing it.

#### Where should the reviewer start?

Start with `WorkerInvocationGuard` in `crates/core/src/plugin/dynamic/worker.rs`, then compare the active-invocation tracking in `crates/worker/src/lib.rs` and `python/plugin/src/nemo_relay_plugin/_api.py`.

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- Closes #334
- Closes RELAY-395

## Summary

* **New Features**
  * Added host-driven cooperative cancellation for unary and streaming worker invocations in both Rust and Python, including timeouts and “stopped consuming” scenarios.
  * Implemented `CancelInvocation` with idempotent, reason-aware acknowledgments.

* **Bug Fixes**
  * Improved cancellation propagation and guaranteed consistent cleanup of in-flight and streamed execution state.
  * Refined cancellation/timeout/inactive error reporting and adjusted timeout messaging granularity.

* **Tests**
  * Added integration and end-to-end coverage for cancellation propagation, dropped futures/streams, and streaming setup edge cases.

* **Documentation**
  * Updated Rust and Python plugin guidance to reflect cooperative async cancellation expectations.

### CODE REVIEW

File: python/plugin/src/nemo_relay_plugin/_api.py [hunk 1]
```diff
@@ -1238,29 +1246,82 @@ async def Register(self, request: Any, context: Any) -> Any:
...
+                with _bind_invocation_scope(request):
+                    stream = await _maybe_await(handler(payload.model_name, llm_request, next_call))
+                    async for value in _as_async_iter(stream):
+                        await queue.put(pb.StreamChunk(value=_json_envelope(JSON_SCHEMA, value)))
+            except asyncio.CancelledError:
+                if active.cancel_reason is not None:
+                    while not queue.empty():
+                        queue.get_nowait()
+                    queue.put_nowait(pb.StreamChunk(error=_cancelled_worker_error(active.cancel_reason)))
+                raise
```
Comment (zhongxuanwang-nv) [inline python/plugin/src/nemo_relay_plugin/_api.py]: A nitty thing — there are cases when `CancelledError` happens but no `cancel_reason` is supplied; however in these cases, this re-raises without emitting a chunk. Then, the consumer at `:1297` hits `if active.task.done() and queue.empty(): guard.finish()` which treats this stream as normal and successful

Suggested fix: in the consumer side, emit a terminal error instead of nothing.

File: python/plugin/src/nemo_relay_plugin/_api.py [hunk 2]
```diff
@@ -1238,29 +1246,82 @@ async def Register(self, request: Any, context: Any) -> Any:
...
-                async for value in _as_async_iter(stream):
-                    yield pb.StreamChunk(value=_json_envelope(JSON_SCHEMA, value))
+            active = self._start_invocation(request.invocation_id, produce())
+            active_ref.append(active)
         except Exception as exc:  # noqa: BLE001 - callback failure is protocol data.
             yield pb.StreamChunk(error=_sdk_error_to_worker(exc))
+            return
+        try:
+            while True:
+                if active.task.done() and queue.empty():
```
Comment (zhongxuanwang-nv) [inline python/plugin/src/nemo_relay_plugin/_api.py:1298]: consumer is here

File: crates/core/src/plugin/dynamic/worker.rs [hunk 3]
```diff
@@ -982,6 +982,83 @@ struct WorkerPluginCallback {
...
+            return;
+        }
+        self.cleaned = true;
+        if !self.continuation_id.is_empty() {
+            self.host_state.remove_continuation(&self.continuation_id);
+        }
+        if !self.scope_stack_id.is_empty() {
+            self.host_state
+                .remove_invocation_scope_stack(&self.scope_stack_id);
+        }
```
Comment (zhongxuanwang-nv) [inline crates/core/src/plugin/dynamic/worker.rs:1051]: A nitty thing, seems like cancellation cleanup removes the invocation's stack alias but it doesn't pop scopes created via` push_scope` nor clears `scope_handles`?

**Review by zhongxuanwang-nv — APPROVED** [review]
Thanks Will! Used the powerful tools to help me on this, and verified these things —

**Review by zhongxuanwang-nv — APPROVED** [review]
